packages/near-membrane-dom/src/browser-realm.ts

```diff
@@ -166,10 +171,103 @@ export default function createVirtualEnvironment(
     if (keepAlive) {
         // TODO: Temporary hack to preserve the document reference in Firefox.
         // https://bugzilla.mozilla.org/show_bug.cgi?id=543435
+        const { document: redDocument } = redWindow;
         ReflectApply(DocumentProtoOpen, redDocument, []);
         ReflectApply(DocumentProtoClose, redDocument, []);
     } else {
         ReflectApply(ElementProtoRemove, iframe, []);
     }
     return env;
 }
+
+function createShadowRealmVirtualEnvironment(
+    globalObjectShape: object | null,
+    globalObjectVirtualizationTarget: WindowProxy & typeof globalThis,
+    providedOptions?: BrowserEnvironmentOptions
+): VirtualEnvironment {
+    if (typeof globalObjectShape !== 'object') {
+        throw new TypeErrorCtor('Missing global object shape.');
+    }
+    if (
+        typeof globalObjectVirtualizationTarget !== 'object' ||
+        globalObjectVirtualizationTarget === null
+    ) {
+        throw new TypeErrorCtor('Missing global object virtualization target.');
+    }
+    const {
+        distortionCallback,
+        endowments,
+        instrumentation,
+        // eslint-disable-next-line prefer-object-spread
+    } = ObjectAssign({ __proto__: null }, providedOptions);
+    if (defaultGlobalPropertyDescriptorMap === null) {
+        const oneTimeIframe = createDetachableIframe();
+        const oneTimeWindow = ReflectApply(
+            HTMLIFrameElementProtoContentWindowGetter,
+            oneTimeIframe,
+            []
+        )!;
+        defaultGlobalOwnKeys = getFilteredGlobalOwnKeys(oneTimeWindow);
+        ReflectApply(ElementProtoRemove, oneTimeIframe, []);
+        defaultGlobalPropertyDescriptorMap = {
+            __proto__: null,
+        } as unknown as PropertyDescriptorMap;
+        assignFilteredGlobalDescriptorsFromPropertyDescriptorMap(
+            defaultGlobalPropertyDescriptorMap,
+            ObjectGetOwnPropertyDescriptors(window)
+        );
+        for (let i = 0, { length } = defaultGlobalOwnKeys; i < length; i += 1) {
+            defaultGlobalOwnKeysRegistry[defaultGlobalOwnKeys[i]] = true;
+        }
+        for (const key in defaultGlobalPropertyDescriptorMap) {
+            if (!(key in defaultGlobalOwnKeysRegistry)) {
+                delete defaultGlobalPropertyDescriptorMap[key];
+            }
+        }
+    }
+    const blueRefs = getCachedGlobalObjectReferences(globalObjectVirtualizationTarget);
+    // Create a new environment.
+    const env = new VirtualEnvironment({
+        blueConnector,
+        distortionCallback,
+        redConnector: createConnector(
+            ReflectApply(FunctionProtoBind, ShadowRealmProtoEvaluate, [new ShadowRealmCtor()])
+        ),
+        instrumentation,
+    });
+    linkIntrinsics(env, globalObjectVirtualizationTarget);
+    // window
+    env.link('globalThis');
```

Comment:
globalThis is not linked before.

Reply:
right... that I'm sure... but on the other side... on the iframe function, I don't see `window` being remapped, hence, I don't know when is this remapped for that to work, that's where the confusion lies for me. Considering that we support browsers where globalThis is not defined... so that might be a problem on itself. There is a `globalThis` in the intrinsics list (since that's a language binding), but again, I just don't know.

Reply:
Oops. I lied. It's linked in `linkIntrinsics`. On the iframe side if globalThis exists linking `window` is skipped as `window === globalThis`.

Reply:
Oops. I lied. It's linked in `linkIntrinsics`. On the iframe side if globalThis exists linking `window` is skipped as `window === globalThis`.

Reply:
I think we should be fine remove this line then! If the browser has ShadowRealm, there will, for sure, be a globalThis binding.

Reply:
Right! Yep. We're safe to remove it.